### ISSUE TITLE
Help with slow open performance with netCDF4 Files

### ISSUE THREAD

**Issue by mathomp4** [Unidata/netcdf-c#489]
### Environment Information
* What platform are you using? (please provide specific distribution/version in summary)
  * [x] Linux
  * [ ] Windows
  * [ ] OSX
  * [ ] Other
  * [ ] NA
* 32 and/or 64 bit?
  * [ ] 32-bit
  * [x] 64-bit
* What build system are you using?
  * [x] autotools (`configure`)
  * [ ] cmake
* Can you provide a sample netCDF file or `C` code to recreate the issue?
  * [x] Yes (please attach to this issue, thank you!)
  * [ ] No
  * [ ] Not at this time

### Summary of Issue
I work on a climate model and I/O is slowly becoming a major bottleneck. By default we try to use nc4 files as our main format (a few Fortran binaries remain, but we are trying to rid ourselves of that). We recently did an experiment to see if our slow performance might be due to the fact that many files we write are written with an unlimited time dimension, so when we read them, they are unlimited. So, we fixed the time dimension, no change. We also compress, chunk, etc. the files but still no big changes. Finally, a colleague suggested converting to netCDF Classic, and, hooboy. Muuuuuuch faster.

Eventually I went whole hog. I took our NC4 file, decompressed it, converted it using `nccopy -k ncX MERR.nc4 MERR.ncX` to nc3, nc5, nc6, nc7 and hacked together a simple program to test it out which you can see here:

https://github.com/mathomp4/netcdfopentester/blob/master/testCNetcdf.c

All it does is open and close a file 1000 times (might be overkill). If I run this on my test files, I get the following output:
```
(1929)(master) $ ./testCNetcdf.exe
nc3 unlimit dim open Time =  0.820 seconds.
nc4 unlimit dim open Time = 19.870 seconds.
nc5 unlimit dim open Time =  1.020 seconds.
nc6 unlimit dim open Time =  1.130 seconds.
nc7 unlimit dim open Time = 20.950 seconds.
nc4 unlimit dim open with hdf5 Time =  0.250 seconds.
```
I added what I _think_ is the correct HDF5 code to open and close a file, but I mainly program netCDF in Fortran and Python, so even this program is a huge success for me! (Also: why it's ugly.) It seems like a netCDF open and an HDF5 open are ~2 orders of magnitude in speed difference. Is there a way to tune netCDF to open like HDF5? Or do you pay for HDF5's quick opening later?

Or, assuming I might have got the HDF5 wrong, why does netcdf Classic open so much faster as well?

Note: The versions of the libraries I'm using are netCDF-C 4.4.1.1 and HDF5 1.8.19. I compiled with Intel 17.0.4 compilers (backed up by GCC 5.3.0). The file is one from the MERRA2 data that can be retrieved (some basic instructions in my README in boring repo mentioned above).

**Comment by edhartnett** [comment]
Can you do an ncdump -s -h of the netCDF-4 file, and show the metadata?

**Comment by mathomp4** [comment]
I can. Here is the file that is the one I uncompressed (MERR.nc4 in the above script).

[MERR.nc4.ncdumphs.txt](https://github.com/Unidata/netcdf-c/files/1318872/MERR.nc4.ncdumphs.txt)

**Comment by DennisHeimbigner** [comment]
Since I do not have the necessary authorization. Can you public post an actual
problematic .nc4 file somewhere temporarily?

**Comment by WardF** [comment]
I'm currently working on a different issue, but I can easily believe that there is an inefficiency somewhere.  Using the `testCNetcdf.exe` above I will, as soon as I can, generate a usage profile and will post the results here for discussion and to see where the overhead is occuring.

**Comment by wkliao** [comment]
Interesting evaluation!
I can confirm the slow open timing when using the files produced from test program "nc_test" which contain a long list of variables and attributes.

% ./read1000 test.nc4
Open test.nc4 1000 times costs 18.308271 seconds
% ./read1000 test.nc5
Open test.nc5 1000 times costs 0.146202 seconds

**Comment by DennisHeimbigner** [comment]
Opening a file with a large metadata is a known problem because, unlike HDF5,
netcdf-4 reads in all the metadata at once, rather than lazily.
However, looking at the file above (MERR.nc4.ncdumphs.txt)
It does not seem to have all that much metadata (~900 lines).
It has about 60 variables and about 800 attributes. I wonder
if our attribute handling code is the problem;  without
looking, I would suspect a linear search problem.

**Comment by DennisHeimbigner** [comment]
After a quick look, it appears that attributes are stored in a singly linked list.
And as each attribute is read, it is stored at the end of the list => linear.
This is the case for all netcdf-4 metadata storage. Looks like it is time
for a comprehensive rebuild of the libsrc4 code to refactor this to use
at least arrays rather than lists. Hash tables would be better.
If memory serves someone did this for the netcdf-3 code already,
so perhaps we can share the code.
In any case, this is a big change, so it is will not occur until at least the next+1
release (e.g. after 4.5.0).

**Comment by WardF** [comment]
I agree. 4.5.1 would be better, 4.5.0 is overdue.

**Comment by mathomp4** [comment]
> Since I do not have the necessary authorization. Can you public post an actual
> problematic .nc4 file somewhere temporarily?

Well, it sounds I'm a little late to the party, but I've put a [compressed version of the file on Dropbox](https://www.dropbox.com/s/cu5qimfb43kpnqg/MERR.compL2.nc4?dl=0). 

> Opening a file with a large metadata is a known problem because, unlike HDF5,
> netcdf-4 reads in all the metadata at once, rather than lazily.

Ah. This makes sense. Although for a bit there, HDF5 opened really fast because I forgot to check the error status (in Fortran). Turns out it'll open very fast if it's just erroring out. But I've confirmed it works. If only I could figure out how to cast the HDF5 file pointer to nf90...best of both worlds!

> After a quick look, it appears that attributes are stored in a singly linked list.
> And as each attribute is read, it is stored at the end of the list => linear.

But it sounds like @DennisHeimbigner might have figured it out. When I get into work tomorrow, I might construct a series of files with varying variables. Luckily I can create a stupid-long file with our climate model and then just remove variables with NCO/CDO. I'd like to know if I can reproduce a linear trend.

**Comment by mathomp4** [comment]
> I'd like to know if I can reproduce a linear trend.

Wrote a quick and dirty tester and...yup:
![netcdf open test](https://user-images.githubusercontent.com/2982494/30694878-f6a12fee-9ea3-11e7-9b4e-ff079d43abcc.png)

The _R<sup>2</sup>_  of the fit didn't export with the image, but it was 0.9871. (https://plot.ly/~Fortran/333)

**Comment by DennisHeimbigner** [comment]
Thanks for the verification.

**Comment by Dave-Allured** [comment]
The virtual attribute _IsNetcdf4 was added in 4.4.1.  I wondered if this was the performance problem, because it requires scanning all the file top metadata.

I compared Netcdf-C 4.4.0/HDF5 1.8.16 to current versions, 4.4.1.1 with 1.10.1.  These timing results were actually run interleaved between versions, to compensate for possible network biases.  I grouped by version and copy/pasted into this summary.

> Netcdf version:
> 4.4.0 of Jan 15 2016 14:13:42 $
> nc4 unlimit dim open Time = 33.390 seconds.
> nc4 unlimit dim open Time = 39.030 seconds.
> nc4 unlimit dim open Time = 36.940 seconds.
> nc4 unlimit dim open Time = 42.890 seconds.
> nc4 unlimit dim open Time = 41.310 seconds.
> nc4 unlimit dim open Time = 43.640 seconds.
> nc4 unlimit dim open Time = 44.250 seconds.
> nc4 unlimit dim open with hdf5 Time =  0.550 seconds. (1.8.16, typical)

> Netcdf version:
> 4.4.1.1 of May 22 2017 10:44:13 $
> nc4 unlimit dim open Time = 51.660 seconds.
> nc4 unlimit dim open Time = 55.440 seconds.
> nc4 unlimit dim open Time = 54.300 seconds.
> nc4 unlimit dim open Time = 56.420 seconds.
> nc4 unlimit dim open Time = 51.510 seconds.
> nc4 unlimit dim open with hdf5 Time =  0.530 seconds. (1.10.1, typical)

This proves that _IsNetcdf4 is not the primary culprit for slow opens, because the attribute was completely absent from netcdf-c 4.4.0.  The main impact is presumably just as Dennis described.  However there seems to be a smaller time increase between versions here.  This might be _IsNetcdf4 or something else.

**Comment by DennisHeimbigner** [comment]
If memory serves, _IsNetcdf4. is computed only when asked for.
So it should not be accessed as part of the above tests.

**Comment by Dave-Allured** [comment]
That makes sense.  Not knowing the internals, I thought it was possible that nc_open might be pre-staging _IsNetcdf4 for some reason.

**Comment by Dave-Allured** [comment]
@mathomp4, your experiment focused on only nc_open.  How much of your model I/O time is consumed in nc_open vs. other file I/O?  There are some known serious inefficiencies when the netcdf-4 chunk layout conflicts with read and write access patterns, or when chunk caches are too small.

My general experience with netcdf-4 is that performance rivals netcdf-3 when the file structure is reasonably matched with the I/O access pattern.  Also in my current model project, 20 to 200 milliseconds is pretty reasonable for a file open that occurs only three times in a 40-minute compute cycle.  Your model's file open pattern might be much different, of course.

**Comment by mathomp4** [comment]
@Dave-Allured Well, this was sort of an odd test. We were looking at I/O and noticed this, so we sort of focused on the open. Now, in most cases, the open probably isn't as important as the read, so we are looking at that too. But we were wondering about open/read speeds when NC4 files were unlimited time v. fixed time, compressed v. uncompressed, etc. But the open was dominating!

We've also encountered what we think is a bug in Intel MPI when reading compressed NC4 files in our model. It was _very_ weird. We found if we uncompressed, the code worked. Now that we have solved the open mystery, we'll probably look at the read more carefully. Chunking, compression, etc. 

But all that said, in one way of running our model we have to open approximately 500 files at start up and ~100 every hour. In that version which was run at very high resolution, I/O (mainly I in this case) was ~25% of the entire model run. Thus, looking at I/O strategies now.

**Comment by Dave-Allured** [comment]
@mathomp4, all things considered, it seems that NC3 format would be a good fit to improve I/O for this particular model.  File opens will be much faster as demonstrated here.  I think data reads and writes will be similar to NC4, possibly faster for some access patterns.  The main thing you give up with NC3 is NC4 data compression.  It seems like speed is your main issue, not storage space.

Watch out for NC3 file structure issues which could bog down data writes.  You should pre-define all variables and attributes in a new NC3 file, before writing any large variables.  Or else utilize the NC3 padding capability.  Otherwise, adding or expanding any variable or attribute can trigger an internal re-write of the whole file when you are not careful.  NC4 is thankfully free of this constraint because of dynamic storage management.  Please excuse me if I am explaining stuff you already know.

If you start now with netcdf-3 format, you should be able to easily upgrade to netcdf-4 if nc_open is improved in a future release.

**Comment by DennisHeimbigner** [comment]
Update: I have been experimenting with the netcdf4 code in 
an attempt to improve the performance of nc_open. I though the
issue was that when the metadata was very large -- multi-megabytes --
the linear searches used by the open() code were causing the slow open speeds.
After running some experiments with highly modified code, I am beginning to believe
that the problem is that the open cost is dominated by the cost for the hdf5 library
to read all the metadata.  Currently, we do not/cannot make use of HDF5
lazy reading of the metadata because we always read in all of the metadata.

So, what can be done. Converting netcdf to fully use lazy metadata read is known
to be a large undertaking. However providing lazy read for attributes seems
to be possible because it appears that most often, attributes are accessed by name.
Does anyone have any insight on this?

**Comment by edhartnett** [comment]
I have (mostly) split the HDF5 code from the libsrc4 data model code. I now see how lazy reads can work for (almost?) all HDF5 objects. I don't think it will be as hard as I previously estimated. 

I deduce from @WardF's recent activity that my HDF4 dispatch PR is soon to be merged, so I can present the HDF5/libsrc4 split after that and we can discuss it. 

After that, it will be an easy step to add lazy reads. Also easy would be to use an array for atts, as we now do for vars, instead of linked lists.

**Comment by DennisHeimbigner** [comment]
After some experimentation,  my current speculation is that the only thing
that needs lazy reading is attributes, and since attributes are mostly read by name,
that should be pretty doable.

**Comment by edhartnett** [comment]
Really? If I have a file with 10000 atts it's slow, but with 10000 vars its not? That seems surprising.

**Comment by DennisHeimbigner** [comment]
I am basing my comment off of the example that others gave me.

**Comment by DennisHeimbigner** [comment]
Also, based on that limited sample. 1000's of variables may not be common.
Our problem is that we do not have a good statistical sample of files.

**Comment by edhartnett** [comment]
OK, but I do agree with you that lazy read of atts would be a good thing.

**Comment by edhartnett** [comment]
What would be great would be to send me the ncdump -h -s of some of the very worst offenders. (Or attach them here.) Pick your three or four slowest opening files, and give me their metadata.

Then I can create some samples and have some real examples to work with.

**Comment by DennisHeimbigner** [comment]
I created some programs to:
1. create a file with large amounts of metadata (parameterized)
2. time the open for a file.
Both are attached
[bigmeta.zip](https://github.com/Unidata/netcdf-c/files/1759996/bigmeta.zip)
We have one other file but it is apparently proprietary.

**Comment by DennisHeimbigner** [comment]
Greg- what might me helpful would be for you to:
1. select some number of such files
2. ncdump -h those files into .cdl files
3. send us those .cdl files.

**Comment by gsjaardema** [comment]
Sorry for the delay.  Here are a couple of bad files.  Writing the testMeshQuad8Psi.e.nc as a netcdf-4 file can take a very long time.

[testMeshQuad8Psi.e.nc.gz](https://github.com/Unidata/netcdf-c/files/1830892/testMeshQuad8Psi.e.nc.gz)
[Block_8192.g.nc.gz](https://github.com/Unidata/netcdf-c/files/1830893/Block_8192.g.nc.gz)

**Comment by mathomp4** [comment]
Query for the netCDF developers. In re this issue, is #857 the one to watch out for for solving this issue?

**Comment by DennisHeimbigner** [comment]
I have made some changes to netcdf4 that improve the performance about 10%.
But the most significant driver for this performance problem is the fact
that netcdf-4 reads all the metadata at once, which is what is being discussed in 
https://github.com/Unidata/netcdf-c/issues/857.
Adding lazy metadata read to netcdf4 is a major undertaking  and we have not decided
yet on the best way to do this.

**Comment by edhartnett** [comment]
I have done some experiments with lazy reads of attributes. Seems very promising and easy to make happen in the code, but I didn't get all tests working.

All this code has changed a lot (with many changes pending as PRs or soon-to-be PRs) so I have to take another stab at this with a clean master.

I suppose the same technique could also be extended to variables.

**Comment by DennisHeimbigner** [comment]
I think the big problem that is going to cause problems with user code is that
lazy operation will change the id numbers of attributes etc. It would be nice
if users did not rely on those numbers, but some of them do.

**Comment by DennisHeimbigner** [comment]
In fact, if memory serves, some of our old test cases even use the numbers as well.

**Comment by edhartnett** [comment]
@gsjaardema when I download and gunzip your test files, the master branch ncdump tells me "Unknown file format." Am I doing something wrong?

**Comment by edhartnett** [comment]
OK, I have a branch which does a lazy read of all attributes.

No att numbers are changed and all tests pass without modification.

This works by only reading the atts (in a group (NC_GLOBAL) or for a variable) when the user first asks for some metadata relating to the attributes.

Since each group and var are still read as a unit, no numbers change. That is, when you ask for any info about any of the attributes of a variable, it will read all variable attributes at that time. If you ask for info about a different attribute of that variable, it now gets it from memory without touching the file.

So the read of attributes is deferred until someone asks for them.

This seems to result in a speedup of about 100% for a file with 100 vars, each with 100 attributes. This is on branch ejh_lazy_atts_2 on the WDC clone. I can prepare a distribution if anyone wants to give it a try.

These changes will go into the queue for merging into master.

**Comment by DennisHeimbigner** [comment]
That is a pretty good compromise. The test_varperf probably is not a good test
for testing this. We need to test it by doing a read of a random number of variables
or groups and then compare again the current eager read code.

**Comment by edhartnett** [comment]
I have a new test, tst_attsperf, which does a test.

It shows a consistent speedup of a factor of 2. That is, it takes half as much time to open a file with lazy atts:
```
ed@mikado:~/tmp/n2/nc_test4$ make tst_attsperf && ./tst_attsperf 
make: 'tst_attsperf' is up to date.
testing speed of open with files with lots of metadata...
Lazy Atts time=76204 Read Atts at Open time=174782 Speedup=2
ok.
*** Tests successful!
```

**Comment by edhartnett** [comment]
I am looking into variables next. I will hit it first thing tomorrow and see what can be done. ;-)

**Comment by edhartnett** [comment]
OK, variables are a harder problem.

There are several difficulties, one is that we probably still want to read all dimensions (i.e. dimscale vars) on open, even if we don't read the vars.

Secondly, everything is processed in a giant loop that touches all HDF5 objects in the file (other than attributes, which are handled separately.) This happens with callback function nc4_rec_read_metadata_cb().

I will continue to look at this. I'm convinced there is a way to do this. I will continue with the libhdf5/libsrc4 separation, and as that happens, I will probably figure out how to do a lazy read of variables. Meanwhile we can get this lazy attribute read into master and that will help.

**Comment by edhartnett** [comment]
OK, I think I see a way to do this for vars as well. @DennisHeimbigner if you could avoid major edits to the libsrc4/libhdf5 directories for a few weeks, I can take a stab at this.

**Comment by DennisHeimbigner** [comment]
Ok, I am mostly working on thredds anyway.

**Comment by gsjaardema** [comment]
@edhartnett The files are the output from ncdump, so no need to run ncdump on them.

**Comment by edhartnett** [comment]
@gsjaardema thanks! Now I see. What a nutty file. This is exactly the kind of thing I never imagined when writing the netcdf-4 code. ;-)

I do have a way forward here, I think. I am working on this now - if I can get it working I will put it up as a PR.

**Comment by edhartnett** [comment]
I have made some good progress here on a new version of the open code which does not read the vars on open.

On open, I open groups and read user-defined types. Vars get a name and varid, but nothing else happens on open. So open is a lot faster now.

As near as I can tell, when the user asks anything about a var in a group, I will have to read all vars in the group, and all vars in all parent groups, up to the root group. I need to do this to catch all the potential dimensions.

This really won't help a lot of the user has 10000 vars in the root group. We will simply delay the reading of all this metadata until the user actually asks for some of it. So nc_open() will be quick but nc_inq() will get slow (the first time it is called).

However, users organizing variables into groups will see a significant improvement. If a group contains variables that are never used in a program, then the metadata for those vars need never be read.

**Comment by edhartnett** [comment]
I wanted to confirm that the recent lazy-att reads really do improve performance, so I did a little testing. Here's some preliminary results that look promising:

![image](https://user-images.githubusercontent.com/4533212/43001616-12ed18e8-8be3-11e8-9039-d4c08913929b.png)

**Comment by DennisHeimbigner** [comment]
Time to break out the perf command.

**Comment by edhartnett** [comment]
OK, after much investigation (including with profiler) I found some bugs in the test code. Also I found that global atts are very much slower than variable atts, and have entered a new ticket to fix that (#1078).

Here's a chart comparing opening a file with lazy atts vs. 4.6.1 vs. HDF5. Now I see that netCDF is just about the same as HDF5, with lazy atts in use. We also see how quickly global atts eat performance for 4.6.1.

![image](https://user-images.githubusercontent.com/4533212/43016659-c5701a56-8c10-11e8-9cef-c71f9cd71a93.png)

**Comment by edhartnett** [comment]
Here are some results with lazy atts for variable attributes. Note the difference in scale - many more attributes are used here because variable atts are so much faster than global atts.

It's not clear why increasing the number of attributes makes the variable slower to read, since variable attributes are not being read. I will investigate.

![image](https://user-images.githubusercontent.com/4533212/43017133-4c2ff3f8-8c12-11e8-9530-13cecb7d2223.png)

**Comment by edhartnett** [comment]
I have just put up a PR with a performance speedup I am calling "Fast Global Att Reads".

I have changed how global atts are read, and this has produced a dramatic speedup. Here's a chart that illustrates:

![image](https://user-images.githubusercontent.com/4533212/43036128-67db214e-8cb8-11e8-9d43-7bf4f0d57342.png)

**Comment by edhartnett** [comment]
Howdy all!

I have been working on an implementation of lazy variable reads. Considerable code changes are required.

Initial results look promising. As we can see from this chart, lazy var reads are about twice as fast, but there is still a strong increase in time with the number of vars in the file. I may be able to reduce this time further - this is just my first attempt. Also I have not yet passed all tests. ;-)

![image](https://user-images.githubusercontent.com/4533212/43322550-1f77e64a-916c-11e8-81b4-12336a9362e5.png)

**Comment by edhartnett** [comment]
OK, I have improved on the above graph a little bit, but I think I've now hit a limitation.

When I do a lazy var read, I actually still open all the HDF5 datasets and create a NC_VAR_INFO_T for them. And that's what is causing the time to be spent. Most of the time is now being spent constructing hashs:

```
                               15745             rehash <cycle 1> [37]
                0.00    0.01   15010/15010       ncindexadd [3]
[15]    16.9    0.00    0.01   30755         NC_hashmapadd <cycle 1> [15]
                0.01    0.00   30755/60780       NC_crc32 [6]
                0.00    0.00   30755/45765       locate [25]
                                  25             rehash <cycle 1> [37]
```

In particular, calculating the crc32 that is used for the hash. Not sure if I can improve that any further. Yet it does seem to be taking a lot of time to generate a mere 5000 hashes.

**Comment by edhartnett** [comment]
I could have shown those profiler results better. Here's the top of the list:

i```
ndex % time    self  children    called     name
                                                 <spontaneous>
[1]     66.7    0.00    0.02                 NC4_process_hdf5_object_cb [1]
                0.00    0.02   15000/15000       nc4_var_list_add2 [2]
                0.00    0.00   15010/15010       ncindexlookup [18]
                0.00    0.00      10/10          nc4_dim_list_add [19]
                0.00    0.00   15000/15000       nc4_var_set_ndims [34]
-----------------------------------------------
```

**Comment by edhartnett** [comment]
Here is an interesting chart that shows the performance of recent netCDF versions, the current master branch, and the current master + fast file opens.

![image](https://user-images.githubusercontent.com/4533212/43339077-9619f15c-9195-11e8-9f59-bb6338eac839.png)

**Comment by edhartnett** [comment]
@DennisHeimbigner here's the full profiler output when I run my test with fast file opens. There are 3000 vars and the test repeats 5 times and gives an average time. So that's why there are 15000 calls to nc4_var_list_add2

```
Flat profile:

Each sample counts as 0.01 seconds.
  %   cumulative   self              self     total           
 time   seconds   seconds    calls  ms/call  ms/call  name    
100.00      0.01     0.01    15025     0.00     0.00  NC_hashmapfree
  0.00      0.01     0.00    60780     0.00     0.00  NC_crc32
  0.00      0.01     0.00    45765     0.00     0.00  locate
  0.00      0.01     0.00    30755     0.00     0.00  NC_hashmapadd
  0.00      0.01     0.00    15077     0.00     0.00  nclistsetalloc
  0.00      0.01     0.00    15052     0.00     0.00  nclistget
  0.00      0.01     0.00    15052     0.00     0.00  nclistnew
  0.00      0.01     0.00    15051     0.00     0.00  nclistfree
  0.00      0.01     0.00    15037     0.00     0.00  nclistpush
  0.00      0.01     0.00    15025     0.00     0.00  NC_hashmapnew
  0.00      0.01     0.00    15025     0.00     0.00  ncindexfree
  0.00      0.01     0.00    15025     0.00     0.00  ncindexnew
  0.00      0.01     0.00    15015     0.00     0.00  NC_hashmapkey
  0.00      0.01     0.00    15010     0.00     0.00  NC_hashmapget
  0.00      0.01     0.00    15010     0.00     0.00  ncindexadd
  0.00      0.01     0.00    15010     0.00     0.00  ncindexith
  0.00      0.01     0.00    15010     0.00     0.00  ncindexlookup
  0.00      0.01     0.00    15000     0.00     0.00  nc4_var_free
  0.00      0.01     0.00    15000     0.00     0.00  nc4_var_list_add2
  0.00      0.01     0.00    15000     0.00     0.00  nc4_var_set_ndims
  0.00      0.01     0.00       47     0.00     0.00  rctrim
  0.00      0.01     0.00       25     0.00     0.00  findPrimeGreaterThan
  0.00      0.01     0.00       25     0.00     0.00  rehash
  0.00      0.01     0.00       16     0.00     0.00  nclistsetlength
  0.00      0.01     0.00       15     0.00     0.00  nclistset
  0.00      0.01     0.00       15     0.00     0.00  obj_track
  0.00      0.01     0.00       10     0.00     0.00  NC_check_id
  0.00      0.01     0.00       10     0.00     0.00  find_in_NCList
  0.00      0.01     0.00       10     0.00     0.00  freestringlist
  0.00      0.01     0.00       10     0.00     0.00  freestringvec
  0.00      0.01     0.00       10     0.00     0.00  nc4_dim_free
  0.00      0.01     0.00       10     0.00     0.00  nc4_dim_list_add
  0.00      0.01     0.00        5     0.00     2.00  NC4_close
  0.00      0.01     0.00        5     0.00     0.00  NC4_get_fileinfo
  0.00      0.01     0.00        5     0.00     0.00  NC4_hdf5get_superblock
  0.00      0.01     0.00        5     0.00     0.00  NC4_open
  0.00      0.01     0.00        5     0.00     0.00  NC_check_file_type
  0.00      0.01     0.00        5     0.00     0.00  NC_interpret_magic_number
  0.00      0.01     0.00        5     0.00     0.00  NC_open
  0.00      0.01     0.00        5     0.00     0.00  NC_urlmodel
  0.00      0.01     0.00        5     0.00     0.00  add_to_NCList
  0.00      0.01     0.00        5     0.00     0.00  del_from_NCList
  0.00      0.01     0.00        5     0.00     0.00  free_NC
  0.00      0.01     0.00        5     0.00     0.00  free_NCList
  0.00      0.01     0.00        5     0.00     0.00  nc4_build_root_grp
  0.00      0.01     0.00        5     0.00     2.00  nc4_close_netcdf4_file
  0.00      0.01     0.00        5     0.00     0.00  nc4_find_nc_file
  0.00      0.01     0.00        5     0.00     0.00  nc4_find_nc_grp_h5
  0.00      0.01     0.00        5     0.00     0.00  nc4_nc4f_list_add
  0.00      0.01     0.00        5     0.00     0.00  nc4_rec_find_grp
  0.00      0.01     0.00        5     0.00     2.00  nc4_rec_grp_del
  0.00      0.01     0.00        5     0.00     2.00  nc_close
  0.00      0.01     0.00        5     0.00     0.00  nc_open
  0.00      0.01     0.00        5     0.00     0.00  nclocate
  0.00      0.01     0.00        5     0.00     0.00  ncuriparse
  0.00      0.01     0.00        5     0.00     0.00  new_NC
  0.00      0.01     0.00        5     0.00     2.00  readfile
  0.00      0.01     0.00        5     0.00     0.00  readmagic
  0.00      0.01     0.00        4     0.00     0.00  rcsearch
  0.00      0.01     0.00        3     0.00     0.00  ncbytessetalloc
  0.00      0.01     0.00        2     0.00     0.00  ncbytesappendn
  0.00      0.01     0.00        2     0.00     0.00  ncloginit
  0.00      0.01     0.00        1     0.00     0.00  NC3_initialize
  0.00      0.01     0.00        1     0.00     0.00  NC4_fileinfo_init
  0.00      0.01     0.00        1     0.00     0.00  NC4_hdf5get_libversion
  0.00      0.01     0.00        1     0.00     0.00  NC4_initialize
  0.00      0.01     0.00        1     0.00     0.00  NCDISPATCH_initialize
  0.00      0.01     0.00        1     0.00     0.00  NC_rcload
  0.00      0.01     0.00        1     0.00     0.00  NC_readfile
  0.00      0.01     0.00        1     0.00     0.00  nc4_hdf5_initialize
  0.00      0.01     0.00        1     0.00     0.00  nc_initialize
  0.00      0.01     0.00        1     0.00     0.00  ncbytesappend
  0.00      0.01     0.00        1     0.00     0.00  ncbytesextract
  0.00      0.01     0.00        1     0.00     0.00  ncbytesfree
  0.00      0.01     0.00        1     0.00     0.00  ncbytesnew
  0.00      0.01     0.00        1     0.00     0.00  ncbytesnull
  0.00      0.01     0.00        1     0.00     0.00  nclog
  0.00      0.01     0.00        1     0.00     0.00  ncsetlogging
  0.00      0.01     0.00        1     0.00     0.00  ncurifree

 %         the percentage of the total running time of the
time       program used by this function.

cumulative a running sum of the number of seconds accounted
 seconds   for by this function and those listed above it.

 self      the number of seconds accounted for by this
seconds    function alone.  This is the major sort for this
           listing.

calls      the number of times this function was invoked, if
           this function is profiled, else blank.

 self      the average number of milliseconds spent in this
ms/call    function per call, if this function is profiled,
	   else blank.

 total     the average number of milliseconds spent in this
ms/call    function and its descendents per call, if this
	   function is profiled, else blank.

name       the name of the function.  This is the minor sort
           for this listing. The index shows the location of
	   the function in the gprof listing. If the index is
	   in parenthesis it shows where it would appear in
	   the gprof listing if it were to be printed.

Copyright (C) 2012-2018 Free Software Foundation, Inc.

Copying and distribution of this file, with or without modification,
are permitted in any medium without royalty provided the copyright
notice and this notice are preserved.

		     Call graph (explanation follows)


granularity: each sample hit covers 2 byte(s) for 100.00% of 0.01 seconds

index % time    self  children    called     name
                0.01    0.00   15025/15025       ncindexfree [2]
[1]    100.0    0.01    0.00   15025         NC_hashmapfree [1]
-----------------------------------------------
                0.00    0.00      25/15025       nc4_rec_grp_del [5]
                0.00    0.01   15000/15025       nc4_var_free [9]
[2]    100.0    0.00    0.01   15025         ncindexfree [2]
                0.01    0.00   15025/15025       NC_hashmapfree [1]
                0.00    0.00   15025/15051       nclistfree [17]
-----------------------------------------------
                0.00    0.01       5/5           nc_close [6]
[3]    100.0    0.00    0.01       5         NC4_close [3]
                0.00    0.01       5/5           nc4_close_netcdf4_file [4]
                0.00    0.00       5/5           nc4_find_nc_grp_h5 [53]
-----------------------------------------------
                0.00    0.01       5/5           NC4_close [3]
[4]    100.0    0.00    0.01       5         nc4_close_netcdf4_file [4]
                0.00    0.01       5/5           nc4_rec_grp_del [5]
                0.00    0.00      15/15051       nclistfree [17]
-----------------------------------------------
                0.00    0.01       5/5           nc4_close_netcdf4_file [4]
[5]    100.0    0.00    0.01       5         nc4_rec_grp_del [5]
                0.00    0.01   15000/15000       nc4_var_free [9]
                0.00    0.00      25/15025       ncindexfree [2]
                0.00    0.00   15010/15010       ncindexith [24]
                0.00    0.00      10/10          nc4_dim_free [38]
-----------------------------------------------
                0.00    0.01       5/5           readfile [7]
[6]    100.0    0.00    0.01       5         nc_close [6]
                0.00    0.01       5/5           NC4_close [3]
                0.00    0.00       5/10          NC_check_id [34]
                0.00    0.00       5/5           free_NC [49]
                0.00    0.00       5/5           del_from_NCList [48]
-----------------------------------------------
                0.00    0.01       5/5           main [8]
[7]    100.0    0.00    0.01       5         readfile [7]
                0.00    0.01       5/5           nc_close [6]
                0.00    0.00       5/5           nc_open [56]
-----------------------------------------------
                                                 <spontaneous>
[8]    100.0    0.00    0.01                 main [8]
                0.00    0.01       5/5           readfile [7]
-----------------------------------------------
                0.00    0.01   15000/15000       nc4_rec_grp_del [5]
[9]     99.8    0.00    0.01   15000         nc4_var_free [9]
                0.00    0.01   15000/15025       ncindexfree [2]
-----------------------------------------------
[10]     0.0    0.00    0.00   15010+15770   <cycle 1 as a whole> [10]
                0.00    0.00   30755             NC_hashmapadd <cycle 1> [13]
                0.00    0.00      25             rehash <cycle 1> [30]
-----------------------------------------------
                0.00    0.00   15010/60780       NC_hashmapget [22]
                0.00    0.00   15015/60780       NC_hashmapkey [21]
                0.00    0.00   30755/60780       NC_hashmapadd <cycle 1> [13]
[11]     0.0    0.00    0.00   60780         NC_crc32 [11]
-----------------------------------------------
                0.00    0.00   15010/45765       NC_hashmapget [22]
                0.00    0.00   30755/45765       NC_hashmapadd <cycle 1> [13]
[12]     0.0    0.00    0.00   45765         locate [12]
-----------------------------------------------
                               15745             rehash <cycle 1> [30]
                0.00    0.00   15010/15010       ncindexadd [23]
[13]     0.0    0.00    0.00   30755         NC_hashmapadd <cycle 1> [13]
                0.00    0.00   30755/60780       NC_crc32 [11]
                0.00    0.00   30755/45765       locate [12]
                                  25             rehash <cycle 1> [30]
-----------------------------------------------
                0.00    0.00      15/15077       nclistset [32]
                0.00    0.00      37/15077       nclistpush [18]
                0.00    0.00   15025/15077       ncindexnew [20]
[14]     0.0    0.00    0.00   15077         nclistsetalloc [14]
-----------------------------------------------
                0.00    0.00       5/15052       nc4_rec_find_grp [55]
                0.00    0.00      10/15052       ncindexlookup [25]
                0.00    0.00      27/15052       NC_rcload [70]
                0.00    0.00   15010/15052       ncindexith [24]
[15]     0.0    0.00    0.00   15052         nclistget [15]
-----------------------------------------------
                0.00    0.00       2/15052       NC_rcload [70]
                0.00    0.00      10/15052       ncuriparse [58]
                0.00    0.00      15/15052       nc4_nc4f_list_add [54]
                0.00    0.00   15025/15052       ncindexnew [20]
[16]     0.0    0.00    0.00   15052         nclistnew [16]
-----------------------------------------------
                0.00    0.00       1/15051       NC_rcload [70]
                0.00    0.00      10/15051       freestringlist [36]
                0.00    0.00      15/15051       nc4_close_netcdf4_file [4]
                0.00    0.00   15025/15051       ncindexfree [2]
[17]     0.0    0.00    0.00   15051         nclistfree [17]
-----------------------------------------------
                0.00    0.00      27/15037       NC_rcload [70]
                0.00    0.00   15010/15037       ncindexadd [23]
[18]     0.0    0.00    0.00   15037         nclistpush [18]
                0.00    0.00      37/15077       nclistsetalloc [14]
-----------------------------------------------
                0.00    0.00   15025/15025       ncindexnew [20]
[19]     0.0    0.00    0.00   15025         NC_hashmapnew [19]
-----------------------------------------------
                0.00    0.00      25/15025       nc4_build_root_grp [51]
                0.00    0.00   15000/15025       nc4_var_list_add2 [26]
[20]     0.0    0.00    0.00   15025         ncindexnew [20]
                0.00    0.00   15025/15052       nclistnew [16]
                0.00    0.00   15025/15077       nclistsetalloc [14]
                0.00    0.00   15025/15025       NC_hashmapnew [19]
-----------------------------------------------
                0.00    0.00       5/15015       nc4_build_root_grp [51]
                0.00    0.00      10/15015       nc4_dim_list_add [39]
                0.00    0.00   15000/15015       nc4_var_list_add2 [26]
[21]     0.0    0.00    0.00   15015         NC_hashmapkey [21]
                0.00    0.00   15015/60780       NC_crc32 [11]
-----------------------------------------------
                0.00    0.00   15010/15010       ncindexlookup [25]
[22]     0.0    0.00    0.00   15010         NC_hashmapget [22]
                0.00    0.00   15010/60780       NC_crc32 [11]
                0.00    0.00   15010/45765       locate [12]
-----------------------------------------------
                0.00    0.00      10/15010       nc4_dim_list_add [39]
                0.00    0.00   15000/15010       nc4_var_list_add2 [26]
[23]     0.0    0.00    0.00   15010         ncindexadd [23]
                0.00    0.00   15010/15010       NC_hashmapadd <cycle 1> [13]
                0.00    0.00   15010/15037       nclistpush [18]
-----------------------------------------------
                0.00    0.00   15010/15010       nc4_rec_grp_del [5]
[24]     0.0    0.00    0.00   15010         ncindexith [24]
                0.00    0.00   15010/15052       nclistget [15]
-----------------------------------------------
                0.00    0.00   15010/15010       NC4_process_hdf5_object_cb [220]
[25]     0.0    0.00    0.00   15010         ncindexlookup [25]
                0.00    0.00   15010/15010       NC_hashmapget [22]
                0.00    0.00      10/15052       nclistget [15]
-----------------------------------------------
                0.00    0.00   15000/15000       NC4_process_hdf5_object_cb [220]
[26]     0.0    0.00    0.00   15000         nc4_var_list_add2 [26]
                0.00    0.00   15000/15015       NC_hashmapkey [21]
                0.00    0.00   15000/15025       ncindexnew [20]
                0.00    0.00   15000/15010       ncindexadd [23]
-----------------------------------------------
                0.00    0.00   15000/15000       NC4_process_hdf5_object_cb [220]
[27]     0.0    0.00    0.00   15000         nc4_var_set_ndims [27]
-----------------------------------------------
                0.00    0.00      47/47          NC_rcload [70]
[28]     0.0    0.00    0.00      47         rctrim [28]
-----------------------------------------------
                0.00    0.00      25/25          rehash <cycle 1> [30]
[29]     0.0    0.00    0.00      25         findPrimeGreaterThan [29]
-----------------------------------------------
                                  25             NC_hashmapadd <cycle 1> [13]
[30]     0.0    0.00    0.00      25         rehash <cycle 1> [30]
                0.00    0.00      25/25          findPrimeGreaterThan [29]
                               15745             NC_hashmapadd <cycle 1> [13]
-----------------------------------------------
                0.00    0.00       1/16          NC_rcload [70]
                0.00    0.00      15/16          nclistset [32]
[31]     0.0    0.00    0.00      16         nclistsetlength [31]
-----------------------------------------------
                0.00    0.00      15/15          obj_track [33]
[32]     0.0    0.00    0.00      15         nclistset [32]
                0.00    0.00      15/15077       nclistsetalloc [14]
                0.00    0.00      15/16          nclistsetlength [31]
-----------------------------------------------
                0.00    0.00       5/15          nc4_build_root_grp [51]
                0.00    0.00      10/15          nc4_dim_list_add [39]
[33]     0.0    0.00    0.00      15         obj_track [33]
                0.00    0.00      15/15          nclistset [32]
-----------------------------------------------
                0.00    0.00       5/10          nc_close [6]
                0.00    0.00       5/10          nc4_find_nc_file [52]
[34]     0.0    0.00    0.00      10         NC_check_id [34]
                0.00    0.00      10/10          find_in_NCList [35]
-----------------------------------------------
                0.00    0.00      10/10          NC_check_id [34]
[35]     0.0    0.00    0.00      10         find_in_NCList [35]
-----------------------------------------------
                0.00    0.00      10/10          ncuriparse [58]
[36]     0.0    0.00    0.00      10         freestringlist [36]
                0.00    0.00      10/15051       nclistfree [17]
-----------------------------------------------
                0.00    0.00      10/10          ncuriparse [58]
[37]     0.0    0.00    0.00      10         freestringvec [37]
-----------------------------------------------
                0.00    0.00      10/10          nc4_rec_grp_del [5]
[38]     0.0    0.00    0.00      10         nc4_dim_free [38]
-----------------------------------------------
                0.00    0.00      10/10          NC4_process_hdf5_object_cb [220]
[39]     0.0    0.00    0.00      10         nc4_dim_list_add [39]
                0.00    0.00      10/15015       NC_hashmapkey [21]
                0.00    0.00      10/15010       ncindexadd [23]
                0.00    0.00      10/15          obj_track [33]
-----------------------------------------------
                0.00    0.00       5/5           NC4_open [42]
[40]     0.0    0.00    0.00       5         NC4_get_fileinfo [40]
                0.00    0.00       5/5           NC4_hdf5get_superblock [41]
-----------------------------------------------
                0.00    0.00       5/5           NC4_get_fileinfo [40]
[41]     0.0    0.00    0.00       5         NC4_hdf5get_superblock [41]
-----------------------------------------------
                0.00    0.00       5/5           NC_open [45]
[42]     0.0    0.00    0.00       5         NC4_open [42]
                0.00    0.00       5/5           nc4_nc4f_list_add [54]
                0.00    0.00       5/5           NC4_get_fileinfo [40]
                0.00    0.00       1/1           nc4_hdf5_initialize [72]
-----------------------------------------------
                0.00    0.00       5/5           NC_open [45]
[43]     0.0    0.00    0.00       5         NC_check_file_type [43]
                0.00    0.00       5/5           readmagic [60]
                0.00    0.00       5/5           NC_interpret_magic_number [44]
-----------------------------------------------
                0.00    0.00       5/5           NC_check_file_type [43]
[44]     0.0    0.00    0.00       5         NC_interpret_magic_number [44]
-----------------------------------------------
                0.00    0.00       5/5           nc_open [56]
[45]     0.0    0.00    0.00       5         NC_open [45]
                0.00    0.00       5/5           new_NC [59]
                0.00    0.00       5/5           add_to_NCList [47]
                0.00    0.00       5/5           NC4_open [42]
                0.00    0.00       5/5           NC_urlmodel [46]
                0.00    0.00       5/5           NC_check_file_type [43]
                0.00    0.00       1/1           nc_initialize [73]
-----------------------------------------------
                0.00    0.00       5/5           NC_open [45]
[46]     0.0    0.00    0.00       5         NC_urlmodel [46]
                0.00    0.00       5/5           ncuriparse [58]
-----------------------------------------------
                0.00    0.00       5/5           NC_open [45]
[47]     0.0    0.00    0.00       5         add_to_NCList [47]
-----------------------------------------------
                0.00    0.00       5/5           nc_close [6]
[48]     0.0    0.00    0.00       5         del_from_NCList [48]
                0.00    0.00       5/5           free_NCList [50]
-----------------------------------------------
                0.00    0.00       5/5           nc_close [6]
[49]     0.0    0.00    0.00       5         free_NC [49]
-----------------------------------------------
                0.00    0.00       5/5           del_from_NCList [48]
[50]     0.0    0.00    0.00       5         free_NCList [50]
-----------------------------------------------
                0.00    0.00       5/5           nc4_nc4f_list_add [54]
[51]     0.0    0.00    0.00       5         nc4_build_root_grp [51]
                0.00    0.00      25/15025       ncindexnew [20]
                0.00    0.00       5/15015       NC_hashmapkey [21]
                0.00    0.00       5/15          obj_track [33]
-----------------------------------------------
                0.00    0.00       5/5           nc4_find_nc_grp_h5 [53]
[52]     0.0    0.00    0.00       5         nc4_find_nc_file [52]
                0.00    0.00       5/10          NC_check_id [34]
-----------------------------------------------
                0.00    0.00       5/5           NC4_close [3]
[53]     0.0    0.00    0.00       5         nc4_find_nc_grp_h5 [53]
                0.00    0.00       5/5           nc4_find_nc_file [52]
                0.00    0.00       5/5           nc4_rec_find_grp [55]
-----------------------------------------------
                0.00    0.00       5/5           NC4_open [42]
[54]     0.0    0.00    0.00       5         nc4_nc4f_list_add [54]
                0.00    0.00      15/15052       nclistnew [16]
                0.00    0.00       5/5           nc4_build_root_grp [51]
-----------------------------------------------
                0.00    0.00       5/5           nc4_find_nc_grp_h5 [53]
[55]     0.0    0.00    0.00       5         nc4_rec_find_grp [55]
                0.00    0.00       5/15052       nclistget [15]
-----------------------------------------------
                0.00    0.00       5/5           readfile [7]
[56]     0.0    0.00    0.00       5         nc_open [56]
                0.00    0.00       5/5           NC_open [45]
-----------------------------------------------
                0.00    0.00       5/5           ncuriparse [58]
[57]     0.0    0.00    0.00       5         nclocate [57]
-----------------------------------------------
                0.00    0.00       5/5           NC_urlmodel [46]
[58]     0.0    0.00    0.00       5         ncuriparse [58]
                0.00    0.00      10/15052       nclistnew [16]
                0.00    0.00      10/10          freestringlist [36]
                0.00    0.00      10/10          freestringvec [37]
                0.00    0.00       5/5           nclocate [57]
-----------------------------------------------
                0.00    0.00       5/5           NC_open [45]
[59]     0.0    0.00    0.00       5         new_NC [59]
-----------------------------------------------
                0.00    0.00       5/5           NC_check_file_type [43]
[60]     0.0    0.00    0.00       5         readmagic [60]
-----------------------------------------------
                0.00    0.00       4/4           NC_rcload [70]
[61]     0.0    0.00    0.00       4         rcsearch [61]
                0.00    0.00       1/1           nclog [79]
-----------------------------------------------
                0.00    0.00       1/3           ncbytesappend [74]
                0.00    0.00       2/3           ncbytesappendn [63]
[62]     0.0    0.00    0.00       3         ncbytessetalloc [62]
-----------------------------------------------
                0.00    0.00       2/2           NC_readfile [71]
[63]     0.0    0.00    0.00       2         ncbytesappendn [63]
                0.00    0.00       2/3           ncbytessetalloc [62]
-----------------------------------------------
                0.00    0.00       1/2           NCDISPATCH_initialize [69]
                0.00    0.00       1/2           nclog [79]
[64]     0.0    0.00    0.00       2         ncloginit [64]
                0.00    0.00       1/1           ncsetlogging [80]
-----------------------------------------------
                0.00    0.00       1/1           nc_initialize [73]
[65]     0.0    0.00    0.00       1         NC3_initialize [65]
-----------------------------------------------
                0.00    0.00       1/1           nc_initialize [73]
[66]     0.0    0.00    0.00       1         NC4_fileinfo_init [66]
                0.00    0.00       1/1           NC4_hdf5get_libversion [67]
-----------------------------------------------
                0.00    0.00       1/1           NC4_fileinfo_init [66]
[67]     0.0    0.00    0.00       1         NC4_hdf5get_libversion [67]
-----------------------------------------------
                0.00    0.00       1/1           nc_initialize [73]
[68]     0.0    0.00    0.00       1         NC4_initialize [68]
-----------------------------------------------
                0.00    0.00       1/1           nc_initialize [73]
[69]     0.0    0.00    0.00       1         NCDISPATCH_initialize [69]
                0.00    0.00       1/2           ncloginit [64]
                0.00    0.00       1/1           NC_rcload [70]
-----------------------------------------------
                0.00    0.00       1/1           NCDISPATCH_initialize [69]
[70]     0.0    0.00    0.00       1         NC_rcload [70]
                0.00    0.00      47/47          rctrim [28]
                0.00    0.00      27/15037       nclistpush [18]
                0.00    0.00      27/15052       nclistget [15]
                0.00    0.00       4/4           rcsearch [61]
                0.00    0.00       2/15052       nclistnew [16]
                0.00    0.00       1/16          nclistsetlength [31]
                0.00    0.00       1/15051       nclistfree [17]
                0.00    0.00       1/1           ncurifree [81]
                0.00    0.00       1/1           ncbytesfree [76]
                0.00    0.00       1/1           ncbytesnew [77]
                0.00    0.00       1/1           NC_readfile [71]
                0.00    0.00       1/1           ncbytesextract [75]
-----------------------------------------------
                0.00    0.00       1/1           NC_rcload [70]
[71]     0.0    0.00    0.00       1         NC_readfile [71]
                0.00    0.00       2/2           ncbytesappendn [63]
                0.00    0.00       1/1           ncbytesnull [78]
-----------------------------------------------
                0.00    0.00       1/1           NC4_open [42]
[72]     0.0    0.00    0.00       1         nc4_hdf5_initialize [72]
-----------------------------------------------
                0.00    0.00       1/1           NC_open [45]
[73]     0.0    0.00    0.00       1         nc_initialize [73]
                0.00    0.00       1/1           NCDISPATCH_initialize [69]
                0.00    0.00       1/1           NC4_initialize [68]
                0.00    0.00       1/1           NC3_initialize [65]
                0.00    0.00       1/1           NC4_fileinfo_init [66]
-----------------------------------------------
                0.00    0.00       1/1           ncbytesnull [78]
[74]     0.0    0.00    0.00       1         ncbytesappend [74]
                0.00    0.00       1/3           ncbytessetalloc [62]
-----------------------------------------------
                0.00    0.00       1/1           NC_rcload [70]
[75]     0.0    0.00    0.00       1         ncbytesextract [75]
-----------------------------------------------
                0.00    0.00       1/1           NC_rcload [70]
[76]     0.0    0.00    0.00       1         ncbytesfree [76]
-----------------------------------------------
                0.00    0.00       1/1           NC_rcload [70]
[77]     0.0    0.00    0.00       1         ncbytesnew [77]
-----------------------------------------------
                0.00    0.00       1/1           NC_readfile [71]
[78]     0.0    0.00    0.00       1         ncbytesnull [78]
                0.00    0.00       1/1           ncbytesappend [74]
-----------------------------------------------
                0.00    0.00       1/1           rcsearch [61]
[79]     0.0    0.00    0.00       1         nclog [79]
                0.00    0.00       1/2           ncloginit [64]
-----------------------------------------------
                0.00    0.00       1/1           ncloginit [64]
[80]     0.0    0.00    0.00       1         ncsetlogging [80]
-----------------------------------------------
                0.00    0.00       1/1           NC_rcload [70]
[81]     0.0    0.00    0.00       1         ncurifree [81]
-----------------------------------------------

 This table describes the call tree of the program, and was sorted by
 the total amount of time spent in each function and its children.

 Each entry in this table consists of several lines.  The line with the
 index number at the left hand margin lists the current function.
 The lines above it list the functions that called this function,
 and the lines below it list the functions this one called.
 This line lists:
     index	A unique number given to each element of the table.
		Index numbers are sorted numerically.
		The index number is printed next to every function name so
		it is easier to look up where the function is in the table.

     % time	This is the percentage of the `total' time that was spent
		in this function and its children.  Note that due to
		different viewpoints, functions excluded by options, etc,
		these numbers will NOT add up to 100%.

     self	This is the total amount of time spent in this function.

     children	This is the total amount of time propagated into this
		function by its children.

     called	This is the number of times the function was called.
		If the function called itself recursively, the number
		only includes non-recursive calls, and is followed by
		a `+' and the number of recursive calls.

     name	The name of the current function.  The index number is
		printed after it.  If the function is a member of a
		cycle, the cycle number is printed between the
		function's name and the index number.


 For the function's parents, the fields have the following meanings:

     self	This is the amount of time that was propagated directly
		from the function into this parent.

     children	This is the amount of time that was propagated from
		the function's children into this parent.

     called	This is the number of times this parent called the
		function `/' the total number of times the function
		was called.  Recursive calls to the function are not
		included in the number after the `/'.

     name	This is the name of the parent.  The parent's index
		number is printed after it.  If the parent is a
		member of a cycle, the cycle number is printed between
		the name and the index number.

 If the parents of the function cannot be determined, the word
 `<spontaneous>' is printed in the `name' field, and all the other
 fields are blank.

 For the function's children, the fields have the following meanings:

     self	This is the amount of time that was propagated directly
		from the child into the function.

     children	This is the amount of time that was propagated from the
		child's children to the function.

     called	This is the number of times the function called
		this child `/' the total number of times the child
		was called.  Recursive calls by the child are not
		listed in the number after the `/'.

     name	This is the name of the child.  The child's index
		number is printed after it.  If the child is a
		member of a cycle, the cycle number is printed
		between the name and the index number.

 If there are any cycles (circles) in the call graph, there is an
 entry for the cycle-as-a-whole.  This entry shows who called the
 cycle (as parents) and the members of the cycle (as children.)
 The `+' recursive calls entry shows the number of function calls that
 were internal to the cycle, and the calls entry for each member shows,
 for that member, how many times it was called from other members of
 the cycle.

Copyright (C) 2012-2018 Free Software Foundation, Inc.

Copying and distribution of this file, with or without modification,
are permitted in any medium without royalty provided the copyright
notice and this notice are preserved.

Index by function name

  [65] NC3_initialize         [50] free_NCList            [23] ncindexadd
   [3] NC4_close              [36] freestringlist          [2] ncindexfree
  [66] NC4_fileinfo_init      [37] freestringvec          [24] ncindexith
  [40] NC4_get_fileinfo       [12] locate                 [25] ncindexlookup
  [67] NC4_hdf5get_libversion [51] nc4_build_root_grp     [20] ncindexnew
  [41] NC4_hdf5get_superblock  [4] nc4_close_netcdf4_file [17] nclistfree
  [68] NC4_initialize         [38] nc4_dim_free           [15] nclistget
  [42] NC4_open               [39] nc4_dim_list_add       [16] nclistnew
  [69] NCDISPATCH_initialize  [52] nc4_find_nc_file       [18] nclistpush
  [43] NC_check_file_type     [53] nc4_find_nc_grp_h5     [32] nclistset
  [34] NC_check_id            [72] nc4_hdf5_initialize    [14] nclistsetalloc
  [11] NC_crc32               [54] nc4_nc4f_list_add      [31] nclistsetlength
  [13] NC_hashmapadd          [55] nc4_rec_find_grp       [57] nclocate
   [1] NC_hashmapfree          [5] nc4_rec_grp_del        [79] nclog
  [22] NC_hashmapget           [9] nc4_var_free           [64] ncloginit
  [21] NC_hashmapkey          [26] nc4_var_list_add2      [80] ncsetlogging
  [19] NC_hashmapnew          [27] nc4_var_set_ndims      [81] ncurifree
  [44] NC_interpret_magic_number [6] nc_close             [58] ncuriparse
  [45] NC_open                [73] nc_initialize          [59] new_NC
  [70] NC_rcload              [56] nc_open                [33] obj_track
  [71] NC_readfile            [74] ncbytesappend          [61] rcsearch
  [46] NC_urlmodel            [63] ncbytesappendn         [28] rctrim
  [47] add_to_NCList          [75] ncbytesextract          [7] readfile
  [48] del_from_NCList        [76] ncbytesfree            [60] readmagic
  [29] findPrimeGreaterThan   [77] ncbytesnew             [30] rehash
  [35] find_in_NCList         [78] ncbytesnull            [10] <cycle 1>
  [49] free_NC                [62] ncbytessetalloc

```

**Comment by DennisHeimbigner** [comment]
What are you using to get these numbers, Linux perf?
In any case, there does not seem to be enough precision
to really see where the time goes. All those zeros give no
information.

**Comment by edhartnett** [comment]
This is gprof, the GNU profiler. 

I will continue to analyze and update when I've got more info.

**Comment by gsjaardema** [comment]
Most or all of my var names are less than 64 chars

**Comment by edhartnett** [comment]
@mathomp4 can you try installing the latest version of HPC netCDF and see how it works for you? It includes all the recent performance improvements for attributes, but does not yet include the lazy var read work.

https://github.com/NOAA-GSD/netcdf-c/releases

**Comment by wkliao** [comment]
Looks like `NC_hashmapfree()` took a significant amount of time.
It mainly consists of a loop calling `free(he->key)`, which implies the `malloc` mechanism has a room for improvement ....?

**Comment by gsjaardema** [comment]
@DennisHeimbigner 
I am not using groups yet other than in some prototypes, so all vars and dims are in the root group.  Some of my complex models have around 500,000 vars

**Comment by gsjaardema** [comment]
I don't think I trust your gprof output.  Gprof is a sampling based profiler and it looks like there aren't enough samples to give you a good indication of what is happening.  I would try using either valgrind with callgrind or perhaps intel vtune or some other profiler that actually counts the true time in a function instead of periodic samples.  

I also don't see any of the utf8 functions in your profile and they tend to show up when I profile my usages. Perhaps the test program is using the library in a way that bypasses the utf8 conversions; if so, that would explain it.

**Comment by edhartnett** [comment]
@gsjaardema the utf8 functions don't show in the profile table because I am only reading, so no utf8 conversions need to be done. 500K variable - well, you are one for the record books. I will try the valgrind profiler next.

@wkliao everything is available in the latest release of HPC NetCDF: https://github.com/NOAA-GSD/netcdf-c/releases. The timing program is nc_test4/tst_attsperf. I did not try classic format files, I will do that for the next round of testing.

@DennisHeimbigner let me rerun tests, as Greg suggests, before you mess with the hashes. Another improvement that is in easy reach (but I have not yet done) is to only create the hash for attributes attributes are actually read. That is, instead of creating a hash table for atts for every var, wait to see if the atts are really going to be read for that var. But first I will rerun the profiiling, as Greg suggests, to ensure we're targeting the correct code for performance improvements.

**Comment by mathomp4** [comment]
@edhartnett I'll need a bit more time. I grabbed it this morning, but it looks like 4.6.2.4 of HPC NetCDF triggers the [same ocprint build error](https://www.unidata.ucar.edu/mailing_lists/archives/netcdfgroup/2018/msg00056.html) on my MacBook as does 4.6.1 of the "standard" NetCDF.

Time to buckle down and try to figure out what's going on with this...

**Comment by edhartnett** [comment]
Try building with --disable-dap.

**Comment by edhartnett** [comment]
Also @mathomp4 is there a github issue for this ocprint error? If not, there should be.

**Comment by mathomp4** [comment]
@edhartnett Well, technically I build with dap on because the ultimate use for this netCDF (the GEOS model and its users) has a need for OpenDAP-enabled netcdf. But for testing...

And I did have a Github issue but as I read the README.md, my reading seemed to me to say that it was more of a mailing list issue. But I can add one.

**Comment by mathomp4** [comment]
Reported as #1095.

**Comment by DennisHeimbigner** [comment]
As an aside, it seems to me that for Greg's example of a single flat group with
500000 variables, the proposed lazy variable mechanism will not help at all.
The moment that one variable is accessed, then all 500000 will be brought in as well.
For his case, we are going to need a true lazy system and that means we need to
be prepared for the effect on variable ids.

**Comment by DennisHeimbigner** [comment]
> instead of creating a hash table for atts for every var, wait to see if the atts are really going to be read for that var. B
I would be surprised if the cost of creating the empty hash table was a significant
factor. It is the insertions that matter, I think.
I have a local branch that is setup to experiment with various name-based
lookup mechanisms (one of which is hashing) and also includes some alternate
hashkey functions. When you are finished, I will attempt to do some perf
measurements to see if any of my alternatives makes much difference.

**Comment by DennisHeimbigner** [comment]
I think the gprof input I would like to see is the profile for a single read
of a metadata file with a relatively large number of variables. This should give
results that do not have so many zeros in them but would more accurately
show what routines are eating the time.

**Comment by edhartnett** [comment]
@DennisHeimbigner I didn't fully explain the lazy vars. Initially I thought that I would have to learn all about all the vars in a group if the user asked for one.

But that is not how the implementatnion works out. Instead, netcdf-4 still iterates through all of the vars in the file on open, but it does not learn all about the vars. (For example, it does not learn their type, or fill value, and some other stuff).

This allows the file to open about twice as fast.

When you ask about a specific variable (for example, what is it's type?) then the library will learn all about that variable, but does not have to learn about all the other variables in the group at that time.

So Greg's file should open about twice as fast with lazy vars.

**Comment by DennisHeimbigner** [comment]
I see. Sounds good.

**Comment by edhartnett** [comment]
@mathomp4 can you try your original file tests with the latest master? Fast global att reads and lazy att reads are working in master, so that might address many of your performance issues.

I am still working on lazy var reads, but they will not be in the upcoming 4.6.2 release.

**Comment by mathomp4** [comment]
@edhartnett I'll give it a try as soon as I can. I'll be at SC18 the next week, so it might be a while (unless the tutorial/workshop/BoF gets really boring!).

**Comment by edhartnett** [comment]
I think with PR #1262 I have reached the end of the effort to speed up file opens. I can't really think of anything else to do to make them faster. ;-)

I will be doing some graphs of the final results, I will post them here before closing this issue.

Those who have been suffering from slow opens are urged to try master after #1262 is merged.

**Comment by mathomp4** [comment]
@edhartnett Is this the "Coordinates" attribute listed in the netCDF docs here: https://www.unidata.ucar.edu/software/netcdf/docs/attribute_conventions.html

I looked at #1262 and I wasn't too sure how I'd enable this in my code. Or is there an example writing a file with COORDINATES in the netCDF source code?

**Comment by edhartnett** [comment]
No, the coordinates attribute is a hidden attribute (called "_Netcdf4Coordinates"). You can see it with h5dump, but not with ncdump, because netCDF hides it. It is not the same as the "Coordinates" attribute you mention.

Once #1262 is merged, there is no need to turn this speedup on, it will be available for all newly created netCDF-4 files. You don't have to write the _NetCDF4Coordinates attribute, the library will do it automatically. 

The presence of _NetCDF4Coordinates will also be detected automatically when reading a file, and it will be used to speed file opens. If it is not present, then files are opened in the usual (slower) way.

**Comment by mathomp4** [comment]
👍 I like speedups/fixes that I don't have to do anything to get! Nice!

**Comment by edhartnett** [comment]
BTW where does the merr sample file come from? That is, what real-world model or application?

**Comment by mathomp4** [comment]
@edhartnett Looking back in time, it seems to be output from MERRA-2, a reanalysis generated by GMAO where I work. Looking at the metadata, it's the tavg1_2d_lnd_Nx aka M2T1NXLND collection on the date 2010-01-05:

https://disc.gsfc.nasa.gov/datasets/M2T1NXLND_V5.12.4/summary

It's 2D land-surface time-averaged output. It's one of the outputs with the larger number of variables:

https://gmao.gsfc.nasa.gov/pubs/docs/Bosilovich785.pdf

which is probably why I looked at it. Though it might have been that I had it at hand when this question/issue came up!
